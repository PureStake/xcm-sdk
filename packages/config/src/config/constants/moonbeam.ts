import { createBalanceBuilder } from '../../balance';
import {
  Assets,
  Chain,
  MOONBEAM_ASSETS_CONFIGS as assets,
  MOONBEAM_CHINS_CONFIGS as chains,
  MoonChain,
  MOON_CHINS_CONFIGS,
} from '../../constants';
import {
  createExtrinsicBuilder,
  XTokensExtrinsicSuccessEvent,
} from '../../extrinsic';
import { MoonbeamAssets } from '../../interfaces';
import { createWithdrawBuilder } from '../../withdraw';
import { ChainXcmConfigs } from '../config.interfaces';

const config = MOON_CHINS_CONFIGS[MoonChain.Moonbeam];
const balance = createBalanceBuilder<MoonbeamAssets>();
const extrinsic = createExtrinsicBuilder<MoonbeamAssets>(config);
const withdraw = createWithdrawBuilder<MoonbeamAssets>();

export const MOONBEAM_CONFIGS: ChainXcmConfigs<MoonbeamAssets> = {
  [Assets.ACA]: {
    asset: assets[Assets.ACA],
    origin: chains[Chain.Acala],
    deposit: [
      {
        balance: balance.system(),
        extrinsic: extrinsic
          .xTokens()
          .transfer()
          .successEvent(XTokensExtrinsicSuccessEvent.TransferredMultiAssets)
          .params(chains[Chain.Acala], {
            Token: assets[Assets.ACA].originSymbol,
          }),
        origin: chains[Chain.Acala],
      },
    ],
    withdraw: [
      withdraw.xTokens({
        balance: balance.system(),
        destination: chains[Chain.Acala],
        existentialDeposit: 100_000_000_000,
        feePerWeight: 8,
      }),
    ],
  },
  [Assets.AUSD]: {
    asset: assets[Assets.AUSD],
    origin: chains[Chain.Acala],
    deposit: [],
    withdraw: [],
  },
  [Assets.DOT]: {
    asset: assets[Assets.DOT],
    origin: chains[Chain.Polkadot],
    deposit: [],
    withdraw: [],
  },
  [Assets.GLMR]: {
    asset: assets[Assets.GLMR],
    origin: config,
    deposit: [],
    withdraw: [],
  },
  [Assets.PARA]: {
    asset: assets[Assets.PARA],
    origin: chains[Chain.Parallel],
    deposit: [],
    withdraw: [],
  },
};
