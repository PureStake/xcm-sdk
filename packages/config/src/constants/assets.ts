export enum Asset {
  ACA = 'ACA',
  ASTR = 'ASTR',
  AUSD = 'AUSD',
  BNC = 'BNC',
  BSX = 'BSX',
  CRAB = 'CRAB',
  CRU = 'CRU',
  CSM = 'CSM',
  DEV = 'DEV',
  DOT = 'DOT',
  GLMR = 'GLMR',
  HKO = 'HKO',
  KAR = 'KAR',
  KBTC = 'KBTC',
  KINT = 'KINT',
  KMA = 'KMA',
  KSM = 'KSM',
  KUSD = 'KUSD',
  LIT = 'LIT',
  MOVR = 'MOVR',
  PARA = 'PARA',
  PARING = 'PARING',
  PHA = 'PHA',
  RMRK = 'RMRK',
  SDN = 'SDN',
  TEER = 'TEER',
  UNIT = 'UNIT',
  USDT = 'USDT',
}
