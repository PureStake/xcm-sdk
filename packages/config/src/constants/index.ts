/* eslint-disable import/no-cycle */
export * from './assets';
export * from './chains';
