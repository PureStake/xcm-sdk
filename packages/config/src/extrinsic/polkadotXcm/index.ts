export * from './polkadotXcm';
export * from './polkadotXcm.interfaces';
