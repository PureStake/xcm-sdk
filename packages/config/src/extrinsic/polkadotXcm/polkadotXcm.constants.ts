export enum PolkadotXcmExtrinsic {
  LimitedReserveTransferAssets = 'limitedReserveTransferAssets',
  LimitedReserveWithdrawAssets = 'limitedReserveWithdrawAssets',
}

export enum PolkadotXcmExtrinsicSuccessEvent {
  Attempted = 'Attempted',
}
