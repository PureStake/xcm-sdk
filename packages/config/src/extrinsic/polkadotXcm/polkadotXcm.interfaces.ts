import { AssetId } from '../../types';
import { Parents } from '../common.interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';

export interface PolkadotXcmPallet {
  pallet: ExtrinsicPallet.PolkadotXcm;
  extrinsic: PolkadotXcmExtrinsic;
  successEvent: PolkadotXcmExtrinsicSuccessEvent;
  getParams: (account: string, amount: bigint) => PolkadotXcmPalletParams;
}

export type PolkadotXcmPalletParams = [
  /**
   * destination
   */
  {
    V1: {
      parents: Parents;
      interior: {
        X1: {
          Parachain: number;
        };
      };
    };
  },
  /**
   * beneficiary
   */
  {
    V1: {
      parents: 0;
      interior: {
        X1: {
          AccountKey20: {
            network: 'Any';
            /**
             * account
             */
            key: string;
          };
        };
      };
    };
  },
  /**
   * asset
   */
  PolkadotXcmAssetParam,
  /**
   * fee
   */
  0,
  /**
   * weight
   */
  {
    Limited: number;
  },
];

export type PolkadotXcmAssetParam =
  | PolkadotXcmAssetParamV0
  | PolkadotXcmAssetParamV1;

export interface PolkadotXcmAssetParamV0 {
  V0: [
    {
      ConcreteFungible: {
        id: 'Null';
        amount: bigint;
      };
    },
  ];
}

export interface PolkadotXcmAssetParamV1 {
  V1: [
    {
      id: {
        Concrete: {
          parents: Parents;
          interior:
            | 'Here'
            | PolkadotXcmAssetParamV1InteriorX1
            | PolkadotXcmAssetParamV1InteriorX2;
        };
      };
      fun: {
        Fungible: bigint;
      };
    },
  ];
}

export interface PolkadotXcmAssetParamV1InteriorX1 {
  X1: {
    PalletInstance: number;
  };
}

export interface PolkadotXcmAssetParamV1InteriorX2 {
  X2:
    | [
        {
          PalletInstance: number;
        },
        {
          GeneralIndex: AssetId;
        },
      ]
    | [
        {
          Parachain: number;
        },
        {
          PalletInstance: number;
        },
      ];
}
