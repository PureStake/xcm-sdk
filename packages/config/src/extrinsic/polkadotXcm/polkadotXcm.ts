import { AssetConfig, ChainConfig, MoonChainConfig } from '../../interfaces';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';
import { getCreateExtrinsic } from './polkadotXcm.util';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function polkadotXcm<Assets>(chain: MoonChainConfig) {
  return {
    limitedReserveTransferAssets: () =>
      limitedReserveTransferAssets<Assets>(chain),
    limitedReserveWithdrawAssets: () => limitedReserveWithdrawAssets(chain),
  };
}

function limitedReserveTransferAssets<Assets>(chain: MoonChainConfig) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveTransferAssets,
          event,
          chain,
          origin,
        );

        return {
          V0: () =>
            createExtrinsic((amount) => ({
              V0: [
                {
                  ConcreteFungible: {
                    id: 'Null',
                    amount,
                  },
                },
              ],
            })),
          V1: () => ({
            here: () =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: 'Here',
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
            X1: () =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X1: {
                            PalletInstance: 5,
                          },
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
            X2: (palletInstance: number, asset: AssetConfig<Assets>) =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 0,
                        interior: {
                          X2: [
                            {
                              PalletInstance: palletInstance,
                            },
                            {
                              // TODO: throw an error if id is not defined or fix types
                              GeneralIndex: asset.originAssetId || 0,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
          }),
        };
      },
    }),
  };
}

function limitedReserveWithdrawAssets(chain: MoonChainConfig) {
  return {
    successEvent: (event: PolkadotXcmExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          PolkadotXcmExtrinsic.LimitedReserveWithdrawAssets,
          event,
          chain,
          origin,
        );

        return {
          V1: () => ({
            X2: (palletInstance: number) =>
              createExtrinsic((amount) => ({
                V1: [
                  {
                    id: {
                      Concrete: {
                        parents: 1,
                        interior: {
                          X2: [
                            {
                              PalletInstance: palletInstance,
                            },
                            {
                              Parachain: chain.parachainId,
                            },
                          ],
                        },
                      },
                    },
                    fun: {
                      Fungible: amount,
                    },
                  },
                ],
              })),
          }),
        };
      },
    }),
  };
}
