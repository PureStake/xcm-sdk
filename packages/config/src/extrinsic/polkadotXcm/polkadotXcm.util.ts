import { ChainKey } from '../../constants';
import { Chain, MoonChain } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  PolkadotXcmExtrinsic,
  PolkadotXcmExtrinsicSuccessEvent,
} from './polkadotXcm.constants';
import {
  PolkadotXcmAssetParam,
  PolkadotXcmPallet,
} from './polkadotXcm.interfaces';

export function getCreateExtrinsic<ChainKeys extends ChainKey>(
  extrinsic: PolkadotXcmExtrinsic,
  event: PolkadotXcmExtrinsicSuccessEvent,
  config: MoonChain,
  origin: Chain<ChainKeys>,
) {
  return (
    getAsset: (amount: bigint) => PolkadotXcmAssetParam,
  ): PolkadotXcmPallet => ({
    pallet: ExtrinsicPallet.PolkadotXcm,
    extrinsic,
    successEvent: event,
    getParams: (account, amount) => [
      {
        V1: {
          parents: 1,
          interior: {
            X1: {
              Parachain: config.parachainId,
            },
          },
        },
      },
      {
        V1: {
          parents: 0,
          interior: {
            X1: {
              AccountKey20: {
                network: 'Any',
                key: account,
              },
            },
          },
        },
      },
      getAsset(amount),
      0,
      {
        Limited: origin.weight,
      },
    ],
  });
}
