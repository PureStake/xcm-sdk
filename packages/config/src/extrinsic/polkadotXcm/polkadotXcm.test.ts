import { Chain, MoonChain } from '../../constants';
import { polkadotXcm } from './polkadotXcm';
import { PolkadotXcmExtrinsicSuccessEvent } from './polkadotXcm.constants';

describe('polkadotXcm', () => {
  const amount = 1000n;
  const account = '0xeF46c7649270C912704fB09B75097f6E32208b85';
  const chain = {
    chain: MoonChain.MoonbaseAlpha,
    name: 'Moonbase Alpha',
    ws: 'wss://wss.api.moonbase.moonbeam.network',
    parachainId: 1000,
  };
  const origin = {
    chain: Chain.AlphanetRelay,
    name: 'Alphanet Relay',
    ws: 'wss://frag-moonbase-relay-rpc-ws.g.moonbase.moonbeam.network',
    weight: 1_000_000_000,
    parachainId: 0,
  };
  const extrinsic = polkadotXcm(chain);

  describe('limitedReserveTransferAssets', () => {
    describe('v0', () => {
      const cfg = extrinsic
        .limitedReserveTransferAssets()
        .successEvent(PolkadotXcmExtrinsicSuccessEvent.Attempted)
        .origin(origin)
        .V0();

      it('should be correct config', () => {
        expect(cfg).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(cfg.getParams(account, amount)).toMatchSnapshot();
      });
    });

    describe('v1', () => {});
  });

  describe('limitedReserveWithdrawAssets', () => {});
});
