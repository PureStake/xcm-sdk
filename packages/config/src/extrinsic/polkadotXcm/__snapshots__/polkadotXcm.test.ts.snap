// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polkadotXcm limitedReserveTransferAssets v0 should be correct config 1`] = `
Object {
  "extrinsic": "limitedReserveTransferAssets",
  "getParams": [Function],
  "pallet": "polkadotXcm",
  "successEvent": "Attempted",
}
`;

exports[`polkadotXcm limitedReserveTransferAssets v0 should get correct params 1`] = `
Array [
  Object {
    "V1": Object {
      "interior": Object {
        "X1": Object {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  Object {
    "V1": Object {
      "interior": Object {
        "X1": Object {
          "AccountKey20": Object {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      },
      "parents": 0,
    },
  },
  Object {
    "V0": Array [
      Object {
        "ConcreteFungible": Object {
          "amount": 1000n,
          "id": "Null",
        },
      },
    ],
  },
  0,
  Object {
    "Limited": 1000000000,
  },
]
`;
